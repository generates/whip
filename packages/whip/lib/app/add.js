export default async function add (...plugins) {
  for (const { plugin, opts } of plugins) {
    // Execute the plugin with the app (this) and any passed options.
    const returned = plugin(this, opts)

    // If the plugin function is a promise, wait for it to resolve.
    if (returned?.then) await returned
  }
}
