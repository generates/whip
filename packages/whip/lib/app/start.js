import http from 'http'
import getHostUrl from '../utilities/getHostUrl.js'

export default function start (port, hostname, callback) {
  const app = this

  // Create the server instance by specifying the app's callback as the handler.
  const server = http.createServer(callback || app.handler)

  // Prefer the port and hostname passed as arguments to those configured in
  // the app even if the port is 0 (which means use a random unused port) so
  // that they can be overridden when serving (e.g. in a test).
  const portToUse = port !== undefined ? port : (app.opts.port || 0)
  const hostnameToUse = hostname || app.opts.hostname
  
  return new Promise((resolve, reject) => {
    server.listen(portToUse, hostnameToUse, function listenCallback (err) {
      if (err) reject(err)

      // Set the server URL (the local URL which can be different from the
      // base URL) so that whatever is starting the server (e.g. tests) can
      // easily know what URL to use.
      server.url = getHostUrl(hostnameToUse, portToUse || server.address().port)

      app.logger.ns('whip.server').info('Server started:', server.url)

      // Add a destroy method to the server instance if not in a production
      // environment (e.g. development or test).
      // https://github.com/nodejs/node/issues/2642
      // Logic influenced by https://github.com/isaacs/server-destroy
      if (!app.opts.isProd) {
        const sockets = []

        // Keep track of all active connections.
        server.on('connection', function storeSocket (socket) {
          const index = sockets.push(socket) - 1
          socket.on('close', () => sockets.splice(index, 1))
        })

        // Add a destroy method to the server instance that closes the server
        // and destroys all active connections.
        server.destroy = function destroy () {
          return new Promise(resolve => server.close(() => {
            for (const socket of sockets) socket.destroy()
            setTimeout(resolve)
          }))
        }
      }

      resolve(server)
    })
  })
}
