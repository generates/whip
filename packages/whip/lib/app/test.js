import { Requester } from '@generates/requester'

const requester = new Requester({ shouldThrow: false })

export default function test (path, opts) {
  const app = this

  // If options has a statusCode, treat it as a response to a previous request
  // and try to extract headers from it to be reused for current request to
  // make it more convenient when writing tests with successive test requests.
  // if (opts?.statusCode) {
  //   const { headers } = opts.request.options
  //   const cookie = opts.headers['set-cookie'] || headers.cookie
  //   const csrf = headers['csrf-token']
  //   opts = { headers: {} }
  //   if (cookie) opts.headers.cookie = cookie
  //   if (csrf) opts.headers['csrf-token'] = csrf
  //   if (app.log) {
  //     app.logger.ns('whip.test').debug(
  //       `Extracted options for ${path} test request`,
  //       opts
  //     )
  //   }
  // }

  return {
    async get () {
      return this.request('get', opts)
    },
    async post (body) {
      return this.requestWithCsrf('post', { ...opts, body })
    },
    async put (body) {
      return this.requestWithCsrf('put', { ...opts, body })
    },
    async delete (body) {
      return this.requestWithCsrf('delete', { ...opts, body })
    },
    async request (method, opts) {
      const server = await app.start(0)
      const response = await requester[method](server.url + path, opts)
      await server.destroy()
      return response
    },
    async requestWithCsrf (method, opts) {
      // const headers = options?.headers
      // if (cfg.test?.csrfPath && (!headers || !headers['csrf-token'])) {
      //   const logger = app.logger && app.logger.ns('whip.test')
      //   if (logger) {
      //     logger.debug(`Adding CSRF token for ${method} ${path} test request`)
      //   }

      //   // Make a request to the CSRF token endpoint to get a CSRF token for
      //   // the test request.
      // const res = await app.test('/', opts).get()

      //   // Add the CSRF token and session cookie to the request headers.
      //   const csrfToken = response.body.csrfToken
      // const cookie = res.headers['set-cookie']
      opts.headers = {
        ...opts.headers
        // ...csrfToken ? { 'csrf-token': csrfToken } : {},
        // ...cookie ? { cookie } : {}
      }

      //   if (logger) {
      //     logger.debug(
      //       `Modified options for ${method} ${path} test request`,
      //       options
      //     )
      //   }
      // }

      return this.request(method, opts)
    }
  }
}
