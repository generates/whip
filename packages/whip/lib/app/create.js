import polka from 'polka'
import { merge } from '@generates/merger'
import add from './add.js'
import start from './start.js'
import test from './test.js'
import logger from '../plugins/logger.js'
import responseMiddleware from '../middleware/response.js'
import requestIdMiddleware from '../middleware/requestId.js'

const defaults = {
  isDev: !process.env.NODE_ENV || process.env.NODE_ENV === 'development',
  isTest: process.env.NODE_ENV === 'test',
  isProd: process.env.NODE_ENV === 'production',
  get hostname () {
    return process.env.APP_HOSTNAME || this.isDev ? 'localhost' : '0.0.0.0'
  },
  port: process.env.PORT
}

export default function create (opts = {}) {
  const app = polka()

  // Add the given options to the app for convenience.
  app.opts = merge({}, defaults, opts)

  /* Methods */

  // Add a method to the app that can be used to add plugins and extend the app.
  app.add = add

  // Add a convenience method to log the server information when listening on a
  // port and serving requests.
  app.start = start

  //
  app.test = test

  /* Plugins */

  // Add a logger to the app and request object.
  app.add({ plugin: logger, opts: app.opts.logger })

  /* Middleware */

  // Add functions to the response object so that sending responses with common
  // formats are more convenient and the response can be logged.
  app.use(responseMiddleware)

  //
  app.use(requestIdMiddleware)

  // Add default onNoMatch handler so that it uses res.send and the response is
  // logged.
  app.onNoMatch = function onNoMatch (req, res) {
    res.statusCode = 404
    res.send('Not Found')
  }

  app.onError = function onError (err, req, res) {
    req.logger.error(err)
    res.statusCode = err.status || 500
    const body = err.body || 'Internal Server Error'
    if (typeof body === 'string') {
      res.send(body)
    } else {
      res.json(body)
    }
  }

  return app
}
