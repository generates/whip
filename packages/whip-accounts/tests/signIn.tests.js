import { test } from '@ianwalter/bff'
import app from './fixtures/app.js'
import { accounts, password } from './fixtures/accounts.js'

const generalUser = accounts.find(a => a.firstName === 'General')
const disabledUser = accounts.find(a => a.firstName === 'Disabled')
const unverifiedUser = accounts.find(a => a.firstName === 'Unverified')

test('Login • No email', async t => {
  const res = await app.test('/sign-in').post({ password })
  t.expect(res.statusCode).toBe(400)
  t.expect(res.body).toMatchSnapshot()
})

test('Login • No password', async t => {
  const res = await app.test('/sign-in').post({ email: generalUser.email })
  t.expect(res.statusCode).toBe(400)
  t.expect(res.body).toMatchSnapshot()
})

test('Login • Invalid credentials', async t => {
  const payload = { ...generalUser, password: 'thisIsNotTheRightPw' }
  const res = await app.test('/sign-in').post(payload)
  t.expect(res.statusCode).toBe(400)
  t.expect(res.body).toEqual({ message: 'Incorrect email or password' })
})

test('Login • Valid credentials', async t => {
  const res = await app.test('/sign-in').post({ ...generalUser, password })
  t.expect(res.statusCode).toBe(201)
  t.expect(res.body).toMatchSnapshot({ csrfToken: t.expect.any(String) })
})

test('Login • Disabled user', async t => {
  const res = await app.test('/sign-in').post({ ...disabledUser, password })
  t.expect(res.statusCode).toBe(400)
  t.expect(res.body).toEqual({ message: 'Incorrect email or password' })
})

test('Login • Already logged in', async t => {
  const credentials = { ...generalUser, password }
  const one = await app.test('/sign-in').post(credentials)
  const two = await app.test('/sign-in', one).post(credentials)
  t.expect(two.statusCode).toBe(201)
  t.expect(two.body).toMatchSnapshot({ csrfToken: t.expect.any(String) })
  t.expect(one.headers['set-cookie']).not.toEqual(two.headers['set-cookie'])
})

test('Login • Unverified user can login', async t => {
  const credentials = { ...unverifiedUser, password }
  const res = await app.test('/sign-in').post(credentials)
  t.expect(res.statusCode).toBe(201)
  t.expect(res.body).toMatchSnapshot({ csrfToken: t.expect.any(String) })
})

test('Login • Remember me', async t => {
  const body = { ...generalUser, password, rememberMe: true }
  let res = await app.test('/sign-in').post(body)
  t.expect(res.statusCode).toBe(201)

  // Sleep for 5 seconds so that the session would expire if rememberMe was
  // false.
  await t.asleep(5000)

  res = await app.test('/account', res).get()
  t.expect(res.statusCode).toBe(200)
})
