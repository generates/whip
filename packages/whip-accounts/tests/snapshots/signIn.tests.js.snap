// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login • No email 1`] = `
Object {
  "feedback": Object {
    "email": Array [
      "A valid email is required.",
    ],
    "username": Array [
      "A valid username is required.",
    ],
  },
  "message": "Validation Error",
}
`;

exports[`Login • No password 1`] = `
Object {
  "feedback": Object {
    "password": Array [
      "A valid password is required.",
    ],
  },
  "message": "Validation Error",
}
`;
